src/ModelManager.ts

```diff
@@ -125,96 +147,181 @@ export class ModelManager {
+     * Fetch page model from store and trigger cq-pagemodel-loaded event
+     * @returns Root page model
+     */
+    private _fetchPageModelFromStore() {
+        const data = this.modelStore.getData();
 
-                                return data;
-                            });
-                        } else {
-                            const data = this.modelStore.getData();
+        triggerPageModelLoaded(data);
 
-                            triggerPageModelLoaded(data);
+        return data;
+    }
 
-                            return data;
+    /**
+     * Sets initialization promise to fetch model if root path is available
+     * Also, to be returned on synchronous initialization
+     */
+    private _setInitializationPromise(rootModelPath: string) {
+        const {
+            rootModelURL
+        } = this._modelPaths;
+
+        this._initPromise = (
+            this._checkDependencies().then(() => {
+                const data = this.modelStore.getData(rootModelPath);
+
+                if (data && (Object.keys(data).length > 0)) {
+                    triggerPageModelLoaded(data);
+
+                    return data;
+                } else if (rootModelURL) {
+                    return this._fetchData(rootModelURL).then((rootModel: Model) => {
+                        try {
+                            this.modelStore.initialize(rootModelPath, rootModel);
+
+                            // If currently active url model isn't available in the stored model, fetch and return it
+                            // If already available, return the root page model from the store
+                            return (
+                                this._fetchActivePageModel(rootModel) || this._fetchPageModelFromStore()
+                            );
+
+                        } catch (e) {
+                            console.error(`Error on initialization - ${e}`);
                         }
-                    } else {
-                        throw new Error('Attempting to retrieve model data from a non-browser. Please provide the initial data with the property key model');
-                    }
-                });
-            }
-        });
+                    });
+                }
+            })
+        );
+    }
 
-        return this._initPromise;
+    /**
+     * Fetch model for the currently active page
+     */
+    private _fetchActivePageModel(rootModel: Model) {
+        const {
+            currentPathname,
+            sanitizedCurrentPathname,
+            metaPropertyModelUrl
+        } = this._modelPaths;
+
+        // Append the child page if the page model doesn't correspond to the URL of the root model
+        // and if the model root path doesn't already contain the child model (asynchronous page load)
+        if (!!currentPathname && !!sanitizedCurrentPathname &&
+            !isPageURLRoot(currentPathname, metaPropertyModelUrl) &&
+            !hasChildOfPath(rootModel, currentPathname)) {
```

Comment:
That's a long condition. It could benefit from an encapsulation + comment

Reply:
There is indeed a preceding comment describing what is covered in this condition. I'll try to make it clearer.